src/printer.js

```diff
@@ -1098,6 +1098,17 @@ function genericPrintNoParens(path, options, print) {
     case "JSXSpreadAttribute":
       return concat([ "{...", path.call(print, "argument"), "}" ]);
     case "JSXExpressionContainer":
+      const shouldIndent =
+        n.expression.type !== 'ArrayExpression' &&
+        n.expression.type !== 'ObjectExpression' &&
+        n.expression.type !== 'ArrowFunctionExpression' &&
+        n.expression.type !== 'CallExpression' &&
+        n.expression.type !== 'FunctionExpression';
+
+      if (!shouldIndent) {
+        return concat(["{", path.call(print, "expression"), "}"]);
+      }
```

Comment:
I'm not sure I get what the `shouldIndent` logic is? Take this example:

```js
<Something>
  {() => (
      <SomethingElse>
        <span />
      </SomethingElse>
    )}
</Something>;
```

I thought that the indent below is the indentation for the opening `{` on the line 2. It's not clear to me why this bug is happening, but it's also getting late so I might not be thinking clearly... Thanks for fixing these JSX bugs, they've annoyed me as well!

I'd like to look at this again tomorrow before merging so I can understand what's happening.